plugins/processors/topk/README.md

```diff
@@ -0,0 +1,97 @@
+# TopK Processor Plugin
+
+The TopK processor plugin is a filter designed to get the top series over a period of time. It can be tweaked to do its top k computation over a period of time, so spikes can be smoothed out.
+
+This plugin groups the metrics based on their name and tags, then generates aggregates values across each group base on fields selected by the user. It then sorts these groups based on these aggregations and returns any metric that belongs to a group in the top k (sorted by any of the aggregations). This means that when calculating the top k, more than k metrics may be returned.
+
+If only the very top k metrics are needed, regardless of grouping, the simple_topk setting will force each metric into its own individual group
+
+### Configuration:
+
+```toml
+[[processors.topk]]
+  ## How many seconds between aggregations
+  # period = 10
+
+  ## How many top metrics to return
+  # k = 10
+
+  ## Metrics are grouped based on their tags and name. The plugin aggregates
+  ## the selected fields of these groups of metrics and sorts the groups based
+  ## these aggregations
+
+  ## Over which tags should the aggregation be done. Globs can be specified, in
+  ## which case any tag matching the glob will aggregated over. If set to an
+  ## empty list is no aggregation over tags is done
+  # group_by = ['*']
+
+  ## Wheter or not to also group by metric name
+  # group_by_metric_name = true
+
+  ## The plugin can aggregate over several fields. If more than one field is
+  ## specified, an aggregation is calculated per group per field.
+
+  ## The plugin returns a metric if it's in a group in the top k groups,
+  ## ordered by any of the aggregations of the selected fields
+
+  ## This effectively means that more than K metrics may be returned. If you
+  ## need to return only the top k metrics regardless of grouping, use the simple_topk setting
+
+
+  ## Over which fields are the top k are calculated
+  # fields = ["value"]
+
+  ## What aggregation to use. Options: sum, mean, min, max
+  # aggregation = "mean"
+
+  ## Instead of the top k largest metrics, return the bottom k lowest metrics
+  # bottomk = false
+
+  ## If true, this will override any GroupBy options and assign each metric
+  ## its own individual group. Default: false
+  # simple_topk = false
```

Comment:
I think the confusion is that I'm thinking of `group_by` having the opposite effect.  In my mind when we `group by *` we are creating buckets for all tag combinations, not grouping together.  I expect it to work this way because of InfluxDB:

```
$ influx
> insert foo,a=x value=42
> insert foo,a=y value=42
> insert foo,a=z value=42
```

If I don't group they are the same series:
```
> select * from foo
name: foo
time                a value
----                - -----
1524523227856932634 x 42
1524523235152945692 y 42
1524523238021967153 z 42
```

If I group by * then I get one series per tag value:
```
> select * from foo group by *
name: foo
tags: a=x
time                value
----                -----
1524523227856932634 42

name: foo
tags: a=y
time                value
----                -----
1524523235152945692 42

name: foo
tags: a=z
time                value
----                -----
1524523238021967153 42
```

Reply:
I expressed myself wrong. I should've say "group by "*" would group metrics that have the same tags with the same values together".

Basically, for grouping, given a list of tags the plugin creates a string of the form `metric_name&tag1=val1&tag2=val2...` for each metric. If a metric doesn't have a tag it doesn't get added to the string, it then places the metric in a map with its generated string as a key.

Reply:
So it actually behaves as you would expect

Reply:
Does that mean that by default you would want `group_by = ["*"]` and that would be identical to `simple_topk`?

Reply:
By default `group_by = ["*"]` its what most people would want I think. But its not the same as `simple_topk`.

The thing to notice is that for `group_by = ["*"]` these metrics belong in the same bucket:
```
foo,a=x, b=y value=42
foo,a=x, b=y value=690
```

But for `simple_topk` they belong in separate buckets. The `simple_topk` option puts every metric in its own bucket **irregladless of tags**, effectively **making `N` buckets each containing a single metric** (where N is the number of ingested metrics)

Reply:
I don't think is is correct, the HashID() is based on measurement name + tagset, they would both have the same ID and they would get aggregated together.

Reply:
No, I feel like it does not behave well within this plugin, as the aggregation aspect is effectively disabled and if you have a period that is different from the input plugin interval you end up with a data set that may be missing some of the intervals completely.

Maybe I shouldn't mention it, but I do wonder if this plugin could be decomposed into several plugins.  The aggregation aspect could be farmed out to the basicstats plugin completely, and then there could be a separate rank aggregator that only ranks based on fields.  Finally there is a processor that filters based on field values (the rank).
```
inputs -> basicstats (or other aggregators) -> ranker aggregator  -> topk processor
```

One thing I like about this is that it separates out the aggregator and processor aspects, so that the plugin is no longer a processor with aggregator like features.  Also if as we add new aggregators you can use them automatically.  Finally, it seems more straightforward to use as a pure annotator, you just leave out the top processor piece.

Reply:
Hi, sorry for the delay, I've had a hectic few days, and also your comment merited some extended thought.

I checked the basicstats plugin and it doesn't have any grouping feature, so I'm guessing that you mean we could port the grouping feature here to basicstats, make a new ranker aggregator, and leave this processor to solely do the filtering.

I like that approach, and I think is better engineered, but I feel that it would delay the plugin an unnecessary amount of time (the old lets-not-release-the-feature-until-is-perfect syndrome)

What I could do is refactor some of the code to "divide it". So I would have a library that would generate the basic stats based on grouping, then another that would generate the ranks, and the processor itself that would do the filtering.

It will delay the release slightly, but that way factoring out things in the future would be significantly simpler, and we get to push this value onto the users sooner. I think is a nice half way to get our cake and eat it too.

What do you think?

Reply:
I think we should probably shelve the idea to split it out, its interesting but I think we shouldn't do it now.  I'm not 100% sure the design would work out, and even if it does I think it would be trickier to setup since you would need to combine so many pieces.  On top of that it would take a fair bit of time to refactor everything.

Let's continue with the current structure on this plugin, and in the future we can experiment more with the modular idea.